src/main/java/com/microsoft/sqlserver/jdbc/SQLServerTimeoutManager.java

```diff
@@ -0,0 +1,139 @@
+import java.util.logging.Logger;
+
+
+/**
+ * Object that handles management of timeout tasks across all sql server connections
+ */
+final class SQLServerTimeoutManager {
+    private static ScheduledExecutorService scheduledTimeoutTasks = Executors.newScheduledThreadPool(1,
+            createThreadFactory("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager"));
+    private static ExecutorService timeoutTaskWorker = Executors.newSingleThreadExecutor(
+            createThreadFactory("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager.TimeoutTaskWorker"));
+    private static Map<UUID, List<TimeoutCommand<?>>> timeoutCommands = new HashMap<>();
+    private final static Logger logger = Logger.getLogger("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager");
+
+    static void startTimeoutCommand(TimeoutCommand<?> timeoutCommand) {
+        if (scheduledTimeoutTasks.isShutdown()) {
+            // reset id counter for timeout commands
+            TimeoutCommand.uniqueId.set(0);
+            scheduledTimeoutTasks = Executors.newScheduledThreadPool(1);
+            timeoutTaskWorker = Executors.newSingleThreadExecutor();
+        }
+
+        timeoutCommand.setTimeoutTask(scheduledTimeoutTasks.schedule(new Runnable() {
+
+            @Override
+            public void run() {
+                Future<?> timeoutTask = timeoutTaskWorker.submit(new Runnable() {
+                    @Override
+                    public void run() {
+                        timeoutCommand.interrupt();
+                    }
+                });
+                try {
+                    // if the timeout command takes too long to interrupt, cancel it and release
+                    timeoutTask.get(10, TimeUnit.SECONDS);
+                } catch (Exception e) {
+                    logger.log(Level.WARNING, "Timeout task too long, aborting", e);
+                } finally {
+                    releaseTimeoutCommand(timeoutCommand);
+                }
+            }
+
+        }, timeoutCommand.getTimeout(), TimeUnit.SECONDS));
+        addTimeoutCommand(timeoutCommand);
+    }
+
+    static void releaseTimeoutCommand(TimeoutCommand<?> timeoutCommand) {
+        removeTimeoutCommand(timeoutCommand);
+        try {
+            if (!timeoutCommand.isTimeoutTaskComplete()) {
+                timeoutCommand.cancelTimeoutTask();
+            }
+            if (!areTimeoutCommandsAvailable()) {
+                scheduledTimeoutTasks.shutdownNow();
+                timeoutTaskWorker.shutdownNow();
+            }
+        } catch (Exception e) {
+            logger.log(Level.WARNING, "Could not cancel timeout task", e);
+        }
+
+    }
+
+    static void releaseTimeoutCommands(UUID connectionId) {
+        synchronized (timeoutCommands) {
+            List<TimeoutCommand<?>> timeouts = timeoutCommands.get(connectionId);
+            if (timeouts != null && !timeouts.isEmpty()) {
+                for (TimeoutCommand<?> timeoutCommand : new ArrayList<>(timeouts)) {

```

Comment:
Also releasing with Iterator of LinkedList will be an O(1) as compared to ArrayList.remove() for each element.